### PR TITLE
CORTX-32221 : Nightly upgrade notification/image-info support

### PR DESCRIPTION
**Pull request by gauravchaudhari02**
# Problem Statement
- Notification for nightly upgrade job is not available.

# Design
-  Added email notification in job.
-  Modified image info to take actual image tag rather than latest tag
-  Changelog image inputs corrected

# Coding
   Checklist for Author
-  [x] Coding conventions are followed and code is consistent

# Testing 
  Checklist for Author
- [ ] Unit and System Tests are added
- [x] Test Cases cover Happy Path, Non-Happy Path and Scalability
- [ ] Testing was performed with RPM
- [x] Jenkins pipelines used for testing - 
   Nightly Upgrade - https://eos-jenkins.colo.seagate.com/job/Cortx-Automation/job/RGW/job/cortx-rgw-cluster-nightly-upgrade/43/
   Upgrade job - https://eos-jenkins.colo.seagate.com/job/Cortx-Automation/job/RGW/job/cortx-rgw-cluster-upgrade/84/
   1-N deployment - https://eos-jenkins.colo.seagate.com/job/Cortx-Automation/job/RGW/job/setup-cortx-rgw-cluster/7986/
    3-N deployment - https://eos-jenkins.colo.seagate.com/job/Cortx-Automation/job/RGW/job/setup-cortx-rgw-cluster/7987/

# Impact Analysis
  Checklist for Author/Reviewer/GateKeeper
- [ ] Interface change (if any) are documented
- [ ] Side effects on other features (deployment/upgrade)
- [ ] Dependencies on other component(s)

# Review Checklist 
  Checklist for Author
- [x] JIRA number/GitHub Issue added to PR
- [x] PR is self reviewed
- [x] Jira and state/status is updated and JIRA is updated with PR link
- [x] Check if the description is clear and explained

# Documentation
  Checklist for Author
- [ ] Changes done to WIKI / Confluence page / Quick Start Guide

### CODE REVIEW

**Review by shailesh-vaidya — COMMENTED** [review]
Please fix Codacy issues. 
Please validate 1N and 3N deployment also.

**Comment by gauravchaudhari02** [comment]
> Please fix Codacy issues. Please validate 1N and 3N deployment also.

@shailesh-vaidya, Ran deployment jobs and links are updated in details.
Regarding codacy issues, looks like we cannot implement changes suggested by codacy. it is failing if we make changes. can you have a look at issues and check if we can avoid/disable those rules?

File: jenkins/automation/kubernetes/upgrade-cortx-cluster.groovy [hunk 1]
```diff
@@ -53,7 +53,7 @@ pipeline {
                 script { build_stage = env.STAGE_NAME }
                 sh label: 'fetch cluster status', script: '''
                     pushd solutions/kubernetes/
-                        echo $hosts | tr ' ' '\n' | head -n 1 > hosts
+                        echo $hosts | tr ' ' '\n' > hosts
```
Comment (shailesh-vaidya) [inline jenkins/automation/kubernetes/upgrade-cortx-cluster.groovy:56]: We need only primary node details for upgrade right so can we continue using `head -n 1`

File: jenkins/internal-ci/generic/nightly/nightly-upgrade.groovy [hunk 2]
```diff
@@ -77,10 +89,77 @@ pipeline {
...
+                env.cortx_script_branch = "${CORTX_SCRIPTS_BRANCH}"
+                env.hosts = sh( script: '''
+                    echo $hosts | tr ' ' '\n' | awk -F["="] '{print $2}'|cut -d',' -f1
+                ''', returnStdout: true).trim()
+                def recipientProvidersClass = [[$class: 'RequesterRecipientProvider']]
+                if ( currentBuild.result == "SUCCESS" ) {
+                    mailRecipients = "CORTX.DevOps.RE@seagate.com"
+                }
+                else {
+                    mailRecipients = "gaurav.chaudhari@seagate.com"
```
Comment (shailesh-vaidya) [inline jenkins/internal-ci/generic/nightly/nightly-upgrade.groovy:141]: Please add `CORTX.DevOps.RE@seagate.com` here also

File: solutions/kubernetes/cortx-upgrade.sh [hunk 3]
```diff
@@ -76,27 +79,39 @@ function check_io_operations() {
...
+    if [ '"$SOLUTION_CONFIG_TYPE"' == "manual" ]; then copy_solution_config '"$REMOTE_SOLUTION_CONFIG"' '"$SCRIPT_LOCATION"'; fi &&
+    add_secondary_separator "Download Upgrade Images on Primary Node" && 
+    CORTX_ACTUAL_SERVER_IMAGE=`pull_image '"$CORTX_SERVER_IMAGE"'` &&
+    echo '"CORTX_ACTUAL_SERVER_IMAGE : \$CORTX_ACTUAL_SERVER_IMAGE"' &&
+    CORTX_ACTUAL_DATA_IMAGE=`pull_image '"$CORTX_DATA_IMAGE"'` &&
+    echo '"CORTX_ACTUAL_DATA_IMAGE : \$CORTX_ACTUAL_DATA_IMAGE"' &&
+    CORTX_ACTUAL_CONTROL_IMAGE=`pull_image '"$CORTX_CONTROL_IMAGE"'` &&
+    echo '"CORTX_ACTUAL_CONTROL_IMAGE : \$CORTX_ACTUAL_CONTROL_IMAGE"' &&
+    add_secondary_separator "Updating CORTX Images info in solution.yaml" &&
+    update_image control-pod '"\$CORTX_ACTUAL_CONTROL_IMAGE"' &&
```
Comment (shailesh-vaidya) [inline solutions/kubernetes/cortx-upgrade.sh:106]: Can we add loop here to minimize code

File: solutions/kubernetes/functions.sh [hunk 4]
```diff
@@ -101,9 +101,19 @@ function nodes_setup() {
 }
 
 function pull_image() {
-    local image=$1
-    add_secondary_separator "Pulling $image image"
-    docker pull $image || { echo "Failed to pull $image"; exit 1; }
+    local image="$1"
+    if [[ "$image" =~ "latest" ]]; then
+        docker pull "$image" &>/dev/null
+        actual_image_tag=$(docker inspect "$image" | grep "VERSION" | grep -o "2.0.0-[0-9][0-9][0-9]*" | head -n 1)
```
Comment (shailesh-vaidya) [inline solutions/kubernetes/functions.sh:107]: We can improve this command as - 
```
[root@ssc-vm-rhev4-0707 ~]# docker inspect --format='{{ index .Config.Cmd}}' $image | grep -o "2.0.0-[0-9][0-9][0-9]*"
2.0.0-835
```

**Review by shailesh-vaidya — CHANGES_REQUESTED** [review]
Please test Jenkins pipelines with the latest code changes.

File: jenkins/automation/kubernetes/upgrade-cortx-cluster.groovy (hunk 1 above)
Comment (gauravchaudhari02) [reply jenkins/automation/kubernetes/upgrade-cortx-cluster.groovy:56]: @shailesh-vaidya, All nodes are needed as provisioner suggested us to download docker images on all nodes so upgrade won't go in timeout stage as sometimes it takes time to download images on worker nodes and reaches time out so upgrade fails.

File: jenkins/internal-ci/generic/nightly/nightly-upgrade.groovy (hunk 2 above)
Comment (gauravchaudhari02) [reply jenkins/internal-ci/generic/nightly/nightly-upgrade.groovy]: Done

File: solutions/kubernetes/cortx-upgrade.sh (hunk 3 above)
Comment (gauravchaudhari02) [reply solutions/kubernetes/cortx-upgrade.sh]: Done

File: solutions/kubernetes/functions.sh (hunk 4 above)
Comment (gauravchaudhari02) [reply solutions/kubernetes/functions.sh]: Done

**Review by shailesh-vaidya — APPROVED** [review]
Approved. 
@mukul-seagate11 , @subhalaxmisa Can you please check and confirm if we can exclude failing checks

**Pull request merged**